### PR TITLE
chore: automate RC environments and regression tests

### PR DESCRIPTION
**Pull request by TomWoodward**
what this does is:

* create unified environment (via unified-deployment scripts) for milestones created in openstax/unified
* when rex-web master is changed:
  * bundle and pre-render the release using books defined in the codebase, name the release `master/<sha>`
  * push it to s3
  * update any environments for milestones that aren't closed with this release
  * find test suites in testrail for environments that were updated and run their test cases in a variety of browsers through saucelabs against the release environment

the pipeline:
image

the overall pipeline is configured in unified-deployment, and deploymenty scripts are there too. the changes here are just for the build/testing scripts that are rex specific

links to cool places:
the pipeline: https://concourse.openstax.org/teams/Dev/pipelines/unified
current testrail plan: https://openstax.testrail.com/index.php?/plans/view/750
unified-deployment pr: https://github.com/openstax/unified-deployment/pull/9

### CODE REVIEW

**Review by m1yag1 — DISMISSED** [review]

File: concourse/create-master-build-configs/script.js
Comment (philschatz) [inline concourse/create-master-build-configs/script.js]: Why is the tag file necessary? It seems like a tag is generated _after_ a commit is created and a tag might never even end up on `master` (e.g. a hotfix but merging it into master might squash the actual commit sha that was on production)

File: concourse/create-master-build-configs/task.yml
Comment (philschatz) [inline concourse/create-master-build-configs/task.yml]: Is it possible to just use `./script.js` (in case any parent directories are renamed)

Hmm, concourse probably does not though. But maybe worth checking?

File: concourse/execute-test-runs/script.bash
Comment (philschatz) [inline concourse/execute-test-runs/script.bash]: Would it be a good time to introduce a shell script linter?

[cnx-recipes](https://github.com/openstax/cnx-recipes) uses [shellcheck](https://github.com/koalaman/shellcheck) which is installed on travis containers

File: concourse/create-master-build-configs/script.js
Comment (TomWoodward) [reply concourse/create-master-build-configs/script.js]: this is the docker tag

File: concourse/create-master-build-configs/task.yml
Comment (TomWoodward) [reply concourse/create-master-build-configs/task.yml]: no, the working directory will be the build folder, this is just the config file it loads

File: concourse/execute-test-runs/script.bash
Comment (TomWoodward) [reply concourse/execute-test-runs/script.bash]: sure, i'm not super familiar with shell scripting best practices, any tools would be helpful.

in general i've been choosing whatever is the easiest language for these scripts, i have some bash, some js, and some ruby. i like that concourse lets you easily choose the best tool for the job but it does make code convention a little messy... open to thoughts...

File: concourse/execute-test-runs/script.bash
Comment (TomWoodward) [inline concourse/execute-test-runs/script.bash]: this isn't the same conditional

File: script/test.bash
Comment (TomWoodward) [inline script/test.bash]: why run `yarn test` from here instead of running this from `yarn test`?

Comment (philschatz) [reply script/test.bash]: It felt a little weird to run shell scripts from `yarn test` but we already run `build-simulations.bash` from it so I can move it there

**Review by philschatz — APPROVED** [review]

File: concourse/execute-test-runs/script.bash
Comment (TomWoodward) [inline concourse/execute-test-runs/script.bash]: @philschatz this internal variable is different after this change. previously it was just the directory name, with the glob now its getting the full path.

why did you change this?

Comment (philschatz) [reply concourse/execute-test-runs/script.bash]: Thanks for testing (or checking the concourse build)! It is probably wrong but I changed it because of the shellcheck message [SC2045: Iterating over ls output is fragile. Use globs](https://github.com/koalaman/shellcheck/wiki/SC2045)

I can add a `# shellcheck disable=SC2045` if there is no good way to write it properly.

https://github.com/koalaman/shellcheck/wiki/Ignore#ignoring-one-specific-instance-in-a-file

Comment (TomWoodward) [reply concourse/execute-test-runs/script.bash]: is there a way to get the glob to only return the directory name instead of the whole path?

Comment (philschatz) [reply concourse/execute-test-runs/script.bash]: hmm, my shell-fu is not great but `basename ${MILESTONE_NUMBER}` might work. How can I test it?

Comment (TomWoodward) [reply concourse/execute-test-runs/script.bash]: https://stackoverflow.com/questions/2664740/extract-file-basename-without-path-and-extension-in-bash might also be relevant to our interests. its hard to test, if you push to this branch concourse will run and "run test plans" should turn green here https://concourse.openstax.org/teams/Dev/pipelines/unified

there is a way to test the job before pushing with `fly execute` but i haven't totally figured out the workflow especially when you have lots of inputs

Comment (philschatz) [reply concourse/execute-test-runs/script.bash]: Thanks! I'll try to make the fixes now that I have a way to verify they work (I forgot that concourse was already running 😊 ).

Also, I got feedback from https://stackoverflow.com/users/26428/dennis-williamson ( @dtwilliamson ) a while back that local variables should be `under_snake` case so they are visually distinct from `ENVIRONMENT_VARS`

Comment (TomWoodward) [reply concourse/execute-test-runs/script.bash]: that makes sense on variable naming

File: concourse/execute-test-runs/script.bash
Comment (TomWoodward) [inline concourse/execute-test-runs/script.bash]: you could use the TR_RUN_PATH here for less string replacing

**Comment by TomWoodward** [comment]
@philschatz your changes are working on concourse.openstax.org, should we wait for @jpslav to review before merging?

**Comment by philschatz** [comment]
👍 for merging now and we can do a concourse discussion when @jpslav gets back from vacation if he wants

**Pull request merged**
